internal/controllers/machineset/machineset_controller.go

```diff
@@ -966,6 +1003,156 @@ func (r *Reconciler) getMachineNode(ctx context.Context, cluster *clusterv1.Clus
+	log := ctrl.LoggerFrom(ctx)
+	// If the version is not defined on the MachineSet then we cannot check for the version skew policies.
+	// Return early.
+	if ms.Spec.Template.Spec.Version == nil {
+		return ctrl.Result{}, nil
+	}
+	msVersion := *ms.Spec.Template.Spec.Version
+
+	cpSemver, err := semver.ParseTolerant(cpVersion)
+	if err != nil {
+		return ctrl.Result{}, errors.Wrapf(err, "failed to perform kubernetes version skew pre-flight check: failed to parse ControlPlane version %s", cpVersion)
+	}
+	msSemver, err := semver.ParseTolerant(msVersion)
+	if err != nil {
+		return ctrl.Result{}, errors.Wrapf(err, "failed to perform kubernetes version skew pre-flight check: failed to parse MachineSet version %s", msVersion)
+	}
+
+	// Check the Kubernetes version skew policy.
+	// => MS minor version cannot be greater than the Control Plane minor version.
+	// => MS minor version cannot be older than 2 minor version of Control Plane.
+	// Kubernetes skew policy: https://kubernetes.io/releases/version-skew-policy/#kubelet
+	floorVersion := semver.Version{
+		Major: cpSemver.Major,
+		Minor: cpSemver.Minor - 2,
+		Patch: 0,
+	}
+	if msSemver.GT(cpSemver) || msSemver.LT(floorVersion) {
+		log.Info(fmt.Sprintf("Pre-flight check failed: MachineSet version (%s) and ControlPlane version (%s) do not conform to the kubernetes version skew policy", msVersion, cpVersion))
+		return ctrl.Result{RequeueAfter: preflightFailedRequeueAfter}, nil
+	}
+
+	return ctrl.Result{}, nil
+}
+
+func (r *Reconciler) kubeadmVersionPreflightCheck(ctx context.Context, cpVersion string, ms *clusterv1.MachineSet) (ctrl.Result, error) {
+	log := ctrl.LoggerFrom(ctx)
+	// If the version is not defined on the MachineSet then we cannot check for the version skew policies.
+	// Return early.
+	if ms.Spec.Template.Spec.Version == nil {
+		return ctrl.Result{}, nil
+	}
+	msVersion := *ms.Spec.Template.Spec.Version
+
+	// If the bootstrap.configRef is nil return early.
+	if ms.Spec.Template.Spec.Bootstrap.ConfigRef == nil {
+		return ctrl.Result{}, nil
+	}
+
+	// If using kubeadm bootstrap provider, check the kubeadm version skew policy.
+	// => MS version should match the Control Plane version.
+	// kubeadm skew policy: https://kubernetes.io/docs/setup/production-environment/tools/kubeadm/create-cluster-kubeadm/#kubeadm-s-skew-against-kubeadm
+	bootstrapConfigRef := ms.Spec.Template.Spec.Bootstrap.ConfigRef
+	groupVersion, err := schema.ParseGroupVersion(bootstrapConfigRef.APIVersion)
+	if err != nil {
+		return ctrl.Result{}, errors.Wrapf(err, "failed to perform kubeadm version skew pre-flight check: failed to parse APIVersion %s", bootstrapConfigRef.APIVersion)
+	}
+	kubeadmBootstrapProviderUsed := bootstrapConfigRef.Kind == "KubeadmConfigTemplate" &&
+		groupVersion.Group == bootstrapv1.GroupVersion.Group
+	if kubeadmBootstrapProviderUsed {
+		if cpVersion != msVersion {
```

Comment:
Should we only compare major+minor version? I think the patch version should not matter

Reply:
(let's move parsing the cpVersion one level up so we do it directly after we get it from the CP and only have to do it once)

Reply:
You are right. The kubeadm version skew only talks about the MINOR. Will change it.